kubernetes/dysk

```diff
@@ -64,106 +63,85 @@ mount() {
   local FS_TYPE=""
                           
   local PAGE_BLOB_PATH=""
-  local VOL_ACCOUNT_NAME=""
 
   ACCOUNTNAME="$(echo  "$@" |"$JQ" -r '.["kubernetes.io/secret/accountname"] // empty' | base64 -d)"
   ACCOUNTKEY="$(echo "$@" | "$JQ" -r '.["kubernetes.io/secret/accountkey"] // empty' | tr -d '\n' | tr -d ' ' | base64 -d)"
-  READWRITE_MODE="$(echo  "$@" | "$JQ" -r '.["kubernetes.io/readwrite"] //empty')"
   FS_TYPE="$(echo  "$@" |  "$JQ" -r '.["kubernetes.io/fsType"] //empty')"
 
-  PAGE_BLOB_PATH="$(echo "$@" | "$JQ" -r '.pageBlobPath //empty')"
-  VOL_ACCOUNT_NAME="$(echo "$@" | "$JQ" -r '.accountName //empty')"
+  CONTAINER_NAME="$(echo "$@" | "$JQ" -r '.container //empty')"
+  BLOB_PAGE_NAME="$(echo "$@" | "$JQ" -r '.blob //empty')"  
+  READWRITE_MODE="$(echo  "$@" | "$JQ" -r '.readwrite //empty')"
+  LEASEID="$(echo  "$@" | "$JQ" -r '.leaseID //empty')"
```

Comment:
Users should not manage lease id directly. This is going to be a very inconvenient UX. Here are the possible scenarios:

**The configmap referenced below has the following fields**
Name:**storageaccount:container:pageblobname**
rw_node_name: **last node that has the dysk as rw mount**

> This config map is used to save the lease id (one per dysk)

### dysk is mounted as *rw* ###

*if no config map exists (matching name == pv.storageaccount:pv.container.pv.pageblobname)*
note: We probably need jitter/retry here. Check ro only mount below.
1. Pass ```--auto-lease``` & ```--break-lease``` args to dyskctl. 
> The assumption here that dysk is new to this cluster.
2. Create configmap and save leaseid returned from dyskctl

*if a configmap exists*

**Case 1: if configmap.rw_node_name == current_node_name**
Get node (configmap.rw_node_name) readiness state:
*if call failed* 
fail the mount and return to kubelet. *this node should no longer have the dysk*

*if success + node is ready*:
#### Mounting ####:
1. mount using existing lease.
2. if failed, break lease, mount and update configmap 

**Case 2: if configmap.rw_node_name != current_node_name**
> This means node with the name ==  configmap.rw_node_name is in a network partition (cluster split brain). Which means something is running and has the disk there or user made a mistake of using dysk as **ReadWriteMany**

#### Process ####
Get node (configmap.rw_node_name) readiness state:
*If call is successful && node is ready** 

Fail because this means dysk is used in ```ReadWriteMany``` mode (or node:configmap.rw_node_name was temporarily in a network split and recovered). The controller will eventually kill one of the pods (on old node or current, does not matter). 

*if node configmap.rw_node_name) is not ready*
1. mount using ```--auto-lease``` && ```--break-lease``` (forcing whatever using the dysk to fail).
2. update existing config map. 

### dysk is mounted as **ro** ###

1. Find an existing configmap with **storageaccount:container:pageblobname** name if found use this lease (arg ```--leaseid``` to dyskctl. if dyskctl responded with bad lease, then follow to step 2
2. Loop + jitter + retry (5 seconds) the following. Try find.. if found use && break, if bad lease continue.
3. If the loop failed to find a configmap with lease, then proceed with mounting as ro with ```--auto-lease``` and ```--break-lease``` args. Then create/update the config map

### unmounting ###
delete the config map **before** unmounting

> The rational behind all this is to support **WriteOnceReadMany** PVCs in a *highly prone to race conditions environment such as clusters*. If you have a better approach that achieves the same please let me know 

On a different note: I am working some short cuts on dyskctl that produces k8s PV directly (check the planning issue).

> a service account token is passed to the flexvol, i believe we can use this to execute any call needed (get config map/node readiness etc..).

Reply:
@khenidak Hi, I first tried to use configMap(see following config), while flex vol driver could only get following info, it could not get configMap info. [configMap](https://kubernetes.io/docs/concepts/storage/volumes/#configmap) data could be consumed by containerized applications **running in a Pod**, while the flex vol driver runs in kubelet, we have difficulty to get configMap inside a driver.

```
[mount /var/lib/kubelet/pods/77654a5d-fcfd-11e7-a11f-000d3afd4b43/volumes/foo~dysk/test {"accountName":"xiazhang3","blob":"dysk4QTmx2HS","container":"dysks","kubernetes.io/fsType":"","kubernetes.io/pod.name":"nginx-flex-dysk-lease","kubernetes.io/pod.namespace":"default","kubernetes.io/pod.uid":"77654a5d-fcfd-11e7-a11f-000d3afd4b43","kubernetes.io/pvOrVolumeName":"test","kubernetes.io/readwrite":"rw","kubernetes.io/secret/accountkey":"...","kubernetes.io/secret/accountname":"eGlhemhhbmcz","kubernetes.io/serviceAccount.name":"default","leaseID":"316ef3f8-255f-4958-b9dd-a8de90c2d40a","readwrite":"ro"}]
```


```
apiVersion: v1
kind: Pod
metadata:
  name: nginx-flex-dysk-lease2
spec:
  containers:
  - name: nginx-flex-dysk-lease
    image: nginx
    env:
      - name: LEASEID
        valueFrom:
          configMapKeyRef:
            name: leaseid
            key: leaseid
    volumeMounts:
    - name: test
      mountPath: /data
  volumes:
  - name: test
    flexVolume:
      driver: "foo/dysk"
      secretRef:
        name: dyskcreds
      options:
        accountName: "xx"
        container: "dysks"
        blob: "dysk4QTmx2HS.vhd"
        leaseID: "316ef3f8-255f-4958-b9dd-a8de90c2d40a"
        readwrite: "ro"
```

Reply:
Hi Andy
The configmap is not to be used by the pod, the pod will be configured as usual with volumes/volume mounts (without leaseid nor reference to config map). 

The configmap will be used by the plugin itself to store leases used by dysk instance. So shell script will execute `kubectl create configmap` `kubectl get configmap` as needed. There is a [service account token](https://github.com/kubernetes/kubernetes/blob/7d716e4e910d8780c0a3b745d637cb63b8169d00/pkg/volume/flexvolume/mounter.go#L70) sent as a parameter, we can use it but this will put us in trouble if the pod is started with service account that has no write access on api server (or pods without service account at all). 

> I am really open to use node's kubelet own kubeconfig to do get/create config maps since pods can be created with service account that has no access to get/create configmaps

Here is a sample pod from a user perspective (using dysk in a pod)
notes:
 
1. The user *does not set* configmap name (the name is *storageaccount:container:pageblobname* or `:` can be `~`)
2. the user *does not set* lease, instead the driver creates/updates leases (saved as config maps, described above).


apiVersion: v1
kind: Pod
metadata:
  name: nginx-flex-dysk-lease2
spec:
  containers:
  - name: nginx-flex-dysk-lease
    image: nginx
    volumeMounts:
    - name: test
      mountPath: /data
  volumes:
  - name: test
    flexVolume:
      driver: **"dysk/dysk"**
      secretRef:
        name: dyskcreds ## This secret has account name and key
        fsType: "ext4"  ## We default to ext4 if this is not provided.
        **readOnly: true** ## Part of flexvol infra, no need to create an option for it 
      options:
        accountName: "xx" ## Validate: options.accountName == SecretRef.AccountName
        container: "dysks"
        blob: "dysk4QTmx2HS.vhd"
        ~leaseID: "316ef3f8-255f-4958-b9dd-a8de90c2d40a"~
        ~readwrite: "ro"~